src/board/audio/sound-emoji-synthesizer.ts

```diff
@@ -0,0 +1,253 @@
+/**
+ * Adapted from Microsoft MakeCode's conversion of the CODAL synthesizer.
+ * Copyright (c) Microsoft Corporation
+ * SPDX-License-Identifier: MIT
+ *
+ * https://github.com/microsoft/pxt/blob/f2476687fe636ec7cbf47e96b22c7acec1978461/pxtsim/sound/soundEmojiSynthesizer.ts
+ */
+
+import { SoundEffect } from "./sound-expressions";
+
+// https://github.com/lancaster-university/codal-microbit-v2/blob/master/inc/SoundEmojiSynthesizer.h#L30
+export const EMOJI_SYNTHESIZER_SAMPLE_RATE = 44100;
+export const EMOJI_SYNTHESIZER_TONE_WIDTH_F = 1024;
+export const EMOJI_SYNTHESIZER_TONE_WIDTH = 1024;
+export const EMOJI_SYNTHESIZER_BUFFER_SIZE = 512;
+
+export const EMOJI_SYNTHESIZER_TONE_EFFECT_PARAMETERS = 2;
+export const EMOJI_SYNTHESIZER_TONE_EFFECTS = 3;
+
+export const EMOJI_SYNTHESIZER_STATUS_ACTIVE = 0x1;
+export const EMOJI_SYNTHESIZER_STATUS_OUTPUT_SILENCE_AS_EMPTY = 0x2;
+export const EMOJI_SYNTHESIZER_STATUS_STOPPING = 0x4;
+
+export class SoundEmojiSynthesizer {
+  bufferSize: number;
+  buffer: number[] = [];
+
+  sampleRate: number;
+  sampleRange: number;
+  samplesPerStep: number[] = [];
+  samplesToWrite: number;
+  samplesWritten: number;
+
+  orMask: number;
+  frequency: number = -1;
+  volume: number;
+
+  position: number;
+  status = 0;
+
+  effectPointer: number = -1;
+  effectBuffer: SoundEffect[] = [];
+
+  get effect() {
+    return this.effectBuffer[this.effectPointer];
+  }
+
+  constructor(
+    id: number,
+    private onDone: () => void,
```

Comment:
I added this onDone callback to mirror DEVICE_SOUND_EMOJI_SYNTHESIZER_EVT_DONE that happens in hardware.